sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -454,6 +492,322 @@ pi_result OCL(piEnqueueMemBufferMap)(
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+// Allocates memory accessible on host and device.
+pi_result OCL(piSharedMemAlloc)(void **result_ptr, pi_context context,
+                                pi_device device,
+                                pi_usm_mem_properties *properties, size_t size,
+                                pi_uint32 alignment) {
+
+  void *Ptr = nullptr;
+  pi_result RetVal = PI_INVALID_OPERATION;
+
+  // First we need to look up the function pointer
+  // It would be good if we could store it for future reuse
+  // Are statics ok?
+  clSharedMemAllocINTEL_fn FuncPtr;
+  RetVal = getExtFuncFromContext<clSharedMemAllocINTEL_fn>(
+      context, "clSharedMemAllocINTEL", &FuncPtr);
+
+  if (RetVal == PI_SUCCESS) {
+    Ptr = FuncPtr(cast<cl_context>(context), cast<cl_device_id>(device),
+                  cast<cl_mem_properties_intel *>(properties), size, alignment,
+                  cast<cl_int *>(&RetVal));
+  }
+
+  *result_ptr = Ptr;
+
+  return RetVal;
+}
+
+pi_result OCL(piMemFree)(pi_context context, void *ptr) {
+
+  clMemFreeINTEL_fn FuncPtr;
+  pi_result RetVal = getExtFuncFromContext<clMemFreeINTEL_fn>(
+      context, "clMemFreeINTEL", &FuncPtr);
+
+  if (RetVal != PI_SUCCESS)
+    return RetVal;
+
+  return cast<pi_result>(FuncPtr(cast<cl_context>(context), ptr));
+}
+
+// Sets up pointer arguments to CL kernels
+pi_result OCL(piKernelSetArgMemPointer)(pi_kernel kernel, pi_uint32 arg_index,
+                                        size_t arg_size,
+                                        const void *arg_value) {
+
+  // Size is unused in CL as pointer args are passed by value.
+
+  // Have to look up the context from the kernel
+  cl_context CLContext;
+  cl_int CLErr = clGetKernelInfo(cast<cl_kernel>(kernel), CL_KERNEL_CONTEXT,
+                                 sizeof(cl_context), &CLContext, nullptr);
+  if (CLErr != CL_SUCCESS) {
+    return cast<pi_result>(CLErr);
+  }
+
+  clSetKernelArgMemPointerINTEL_fn FuncPtr;
+  pi_result RetVal = getExtFuncFromContext<clSetKernelArgMemPointerINTEL_fn>(
+      cast<pi_context>(CLContext), "clSetKernelArgMemPointerINTEL", &FuncPtr);
+
+  if (RetVal != PI_SUCCESS)
+    return RetVal;
+
+  // OpenCL passes pointers by value not by reference
+  // This means we need to deref the arg to get the pointer value
+  auto PtrToPtr = reinterpret_cast<const intptr_t *>(arg_value);
+  auto DerefPtr = reinterpret_cast<void *>(*PtrToPtr);
+  return cast<pi_result>(FuncPtr(cast<cl_kernel>(kernel), arg_index, DerefPtr));
+}
+
+// Enables indirect access of pointers in kernels.
+// Necessary to avoid telling CL about every pointer that might be used.
+pi_result OCL(piKernelSetIndirectAccess)(pi_kernel kernel, pi_queue queue) {
```

Comment:
instead of adding this new interface could we rather make it a new option of piKernelCreate to specify this behavior?

Reply:
If you wanted to do that, we'd have to make a special version of piEnqueueKernelLaunch to check this flag and do the right thing.

Reply:
I see that we don't have any configuration parameter for kernel neither at creation nor at submission. So I agree that we want a new interface to set kernel attributes. Can we make this more generic though? e.g. piextKernelSetAttribute?

Reply:
That's a bit separate, but you're basically arguing for wiring up clSetKernelExecInfo.